synapse/http/server.py

```diff
@@ -254,7 +209,89 @@ def register_paths(self, method, path_patterns, callback):
         pass
 
 
-class JsonResource(HttpServer, resource.Resource):
+class _AsyncResource(resource.Resource, metaclass=abc.ABCMeta):
+    """Base class for resources that have async handlers.
+    """
+
+    def render(self, request):
+        """ This gets called by twisted every time someone sends us a request.
+        """
+        defer.ensureDeferred(self._async_render(request))
+        return NOT_DONE_YET
+
+    @wrap_async_request_handler
+    async def _async_render(self, request):
+        """ This gets called from render() every time someone sends us a request.
+
+        Calls `self._get_handler_for_request` to get the callback to use.
+        """
+        try:
+            callback, servlet_classname, group_dict = self._get_handler_for_request(
+                request
+            )
+
+            # Make sure we have a name for this handler in prometheus.
+            request.request_metrics.name = servlet_classname
+
+            # Now trigger the callback. If it returns a response, we send it
+            # here. If it throws an exception, that is handled by the wrapper
+            # installed by @request_handler.
+            kwargs = intern_dict(
+                {
+                    name: urllib.parse.unquote(value) if value else value
+                    for name, value in group_dict.items()
+                }
+            )
+
+            raw_callback_return = callback(request, **kwargs)
+
+            # Is it synchronous? We'll allow this for now.
+            if isinstance(raw_callback_return, (defer.Deferred, types.CoroutineType)):
+                callback_return = await raw_callback_return
+            else:
+                callback_return = raw_callback_return
+
+            if callback_return is not None:
+                code, response = callback_return
+                self._send_response(request, code, response)
+        except Exception:
+            f = failure.Failure()
```

Comment:
probably needs a comment to stop someone "optimising" this to just use the caught exception